### DISCUSSION TITLE
Tell us about your Moquette usage

### DISCUSSION THREAD

**Discussion by andsel** [moquette-io/moquette#874] [General]
Hi folks, I open this discussion to know from you, which is your experience with Moquette MQTT broker.
How do you use it, as standalone broker or inside a product? Tell us, if you can, what strengts you have discovered and which are the pain points or the things that can be improved.
A couple of existing feedbacks:
- #719
- #684

I would love to know from the peoples that uses it, your stories.

**Comment by robinwu2008** [comment]
I use mqtt to push messages to the terminal so need mqtt server to support personalized authentication like reading the user's token from mysql or redis

**Comment by robinwu2008** [comment]
I also need to use mqtt as a im server, so I expect mqtt server to provide some interfaces for message persistence, offline messages, etc

**Comment by SpringStudent** [comment]
I use this project as a communication service for the RPC-MQTT middleware. In our IoT domain, the public server needs to invoke a specific method on a client within the local area network. In this case, Moquette acts as a data relay station for data forwarding.
here is an demo project https://github.com/SpringStudent/rpc-mqtt

**Comment by Lilithsy** [comment]
从上个月开始一直抛出这个错误是不是跟更新依赖有关呢03-27 11:06:49.421 ERROR[oup-6-4]broker.NewNettyMQTTHandler:Unexpected exception while processing MQTT message. Closing Netty channel. CId=null

java.net.SocketException: Connection reset
        at sun.nio.ch.SocketChannelImpl.throwConnectionReset(SocketChannelImpl.java:394) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:426) ~[?:?]
        at io.netty.buffer.PooledByteBuf.setBytes(PooledByteBuf.java:255) ~[netty-buffer-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1132) ~[netty-buffer-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:356) ~[netty-transport-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:151) ~[netty-transport-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796) ~[netty-transport-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732) ~[netty-transport-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658) ~[netty-transport-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998) ~[netty-common-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.119.Final.jar!/:4.1.119.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.119.Final.jar!/:4.1.119.Final]
        at java.lang.Thread.run(Thread.java:833) [?:?]怎么解决这个问题

**Reply by andsel** [reply]
Hi @Lilithsy this discussion is to tell us your experience with Moquette and in which field/context you use it.
It's not intended for bug report, please file a new bug report under https://github.com/moquette-io/moquette/issues describing the problem you face in English please.

**Reply by Lilithsy** [reply]
Sent a question but did not reply, please reply. Thank you

**Comment by hm-zeroone** [comment]
Hi Moquette team,

We’re building Orbit OS, a new lightweight embedded Linux platform that works like an Android-style app store for remote deployment and management of apps and services.

https://www.orbit-os.org

We really like Moquette MQTT Broker and think it’s a great fit for IoT and edge devices.

We’d like to explore the possibility of officially publishing Moquette as a maintained app in our Orbit OS store:
https://store.orbit-os.org

This would enable simple one-click remotely deployment on embedded devices and increase visibility in IoT/edge environments.

If you're interested, we’d be happy to collaborate and help with packaging.

Thanks!

**Reply by andsel** [reply]
Hi @HugoMSMartins,
thanks for the appreciation of Moquette project. Woudl be great to have Moquette as an available choice on Orbit OS, but don't have time to spend to learn and maintain that. I can help with specific problems on Moquette, but can sustain the ownership of such effort ATM. I'm more than happy to help you on specific bug/issue/explanation on Moquette side.

Cheers :-)

**Reply by hm-zeroone** [reply]
Hi @andsel,

Thanks a lot for the quick reply — totally understandable.

We’ll move forward with packaging and maintaining the app on our side.  
Really appreciate your openness to help on Moquette-related topics if needed.

Just one quick question — would you be OK with us using the Moquette name and icon for the app listing in Orbit OS, with clear attribution to the project and your team?

We want to make sure we represent the project properly and respectfully.

Looking forward to bringing Moquette to Orbit OS 🙂

Thanks again!

**Reply by andsel** [reply]
> would you be OK with us using the Moquette name and icon for the app listing in Orbit OS, with clear attribution to the project and your team?

Yep go ahead, no problem. You can find the logo at: https://github.com/moquette-io/moquette/blob/gh-pages/images/Moquette-LOGO-Background-DARK_200px.png

**Reply by hm-zeroone** [reply]
Hi @andsel,

Thanks a lot — really appreciate it!

We’ll make sure to properly attribute the project and use the logo accordingly.

We’ll share the link here as soon as the app is published on Orbit OS store.

Looking forward to bringing Moquette to Orbit OS 🙂

Thanks again for your support!

**Reply by hm-zeroone** [reply]
Hi @andsel,

Following up on your OK to use the Moquette name and logo — the OrbitOS app is live now:

[https://store.orbit-os.org/app/moquette-jar](https://store.orbit-os.org/app/moquette-jar)

Thanks again for the quick yes, and for building Moquette in the first place — it's been a great fit for running an embedded broker on these devices.

A few things, whenever you have a moment:

**Feedback on the integration** — if you'd like to take a look, I'd really value your read on whether anything looks misconfigured, or just not in the spirit of how Moquette is meant to be used.
**Linking back** — would it make sense for the Moquette project to link to the OrbitOS app as an embedded-devices use case? Happy to go whichever way you'd prefer.
**Developer credit** — in the store listing, is it OK if we list the developer as "Moquette team"? Or would you rather we credit it some other way — or not tie it to "Moquette team" at all?

This is our first version — simple and functional. Moquette supports quite a few more configuration options than we expose in the web admin today; we'll be adding those over time.

**Comment by AAA123369** [comment]
/**
     * 通知拦截器：有客户端发布了 MQTT 消息
     * <p>
     * 【ByteBuf 引用計数管理说明】
     * Netty 的 ByteBuf 使用引用计数（refCount）管理底层直接内存（Direct Memory）的生命周期。
     * - refCount > 0：内存有效，可安全访问
     * - refCount = 0：内存已释放，访问将导致 UseCntException
     * - GC 只能回收 Java 对象（ByteBuf 壳），但不会调用 release()，因此底层直接内存不会回池
     * - 必须显式调用 release() 才能将内存归还给 Netty 内存池
     * <p>
     * 【本方法的引用计数流转】
     * 1. 调用方传入 msg，refCount = N（由 Netty Pipeline 管理）
     * 2. msg.retain()：refCount = N+1，确保异步执行期间内存不被 Pipeline 释放
     * 3. 每个 handler 收到 msg.retainedDuplicate()：创建独立副本（refCount = N+2），用完后必须 release
     * 4. lambda 结束前 release(msg)：refCount 回到 N+1 → Pipeline 后续释放回到 N
     * <p>
     * 【修复说明 - Moquette 0.17 ByteBuf 泄漏缺陷】
     * 原始代码：handler.onPublish(new InterceptPublishMessage(msg.retainedDuplicate(), ...))
     * 问题：retainedDuplicate() 创建的副本从未被 release()，导致每次消息处理泄漏一个 ByteBuf 的直接内存。
     * 修复：在 handler 处理完成后，通过 finally 块确保副本被释放。
     *
     * @param msg      MQTT 发布消息（包含 topic、payload 等，底层为引用计数的 ByteBuf）
     * @param clientID 发布消息的客户端 ID
     * @param username 发布消息的用户名
     */
    @Override
    public void notifyTopicPublished(final MqttPublishMessage msg, final String clientID, final String username) {
        // 增加引用计数，防止消息在异步执行期间被 Netty Pipeline 释放
        // retain() 后 refCount = N+1，对应的 release() 在 lambda 的 finally 块中
        msg.retain();

        try {
            // 将拦截器通知任务提交到线程池异步执行
            executor.execute(() -> {
                try {
                    // 提取消息元数据（topic 和 messageId 为 String/int 类型，不涉及 ByteBuf 引用）
                    int messageId = msg.variableHeader().messageId();
                    String topic = msg.variableHeader().topicName();

                    // 遍历所有注册了 PUBLISH 事件的拦截器，逐一通知
                    for (InterceptHandler handler : handlers.get(InterceptPublishMessage.class)) {
                        LOG.debug("Notifying MQTT PUBLISH message to interceptor. CId={}, messageId={}, topic={}, "
                                + "interceptorId={}", clientID, messageId, topic, handler.getID());

                        // retainedDuplicate() 创建一个共享底层数据的独立副本：
                        // - 副本拥有独立的 readerIndex/writerIndex
                        // - 底层 ByteBuf 的 refCount +1（因此用完后必须 release）
                        MqttPublishMessage dup = msg.retainedDuplicate();
                        try {
                            // 将副本传递给拦截器处理
                            // 拦截器可读取消息内容，但不应持有或延迟释放该副本
                            handler.onPublish(new InterceptPublishMessage(dup, clientID, username));
                        } finally {
                            // 【关键修复】无论 handler 是否抛异常，都释放副本的直接内存
                            // 不释放会导致 ByteBuf 泄漏：GC 只回收 Java 对象，底层内存永不回池
                            ReferenceCountUtil.release(dup);
                        }
                    }
                } finally {
                    // 释放 retain() 增加的引用计数，与第 125 行的 msg.retain() 配对
                    // 此时 refCount 从 N+1 降回 N，由 Netty Pipeline 后续正常管理
                    ReferenceCountUtil.release(msg);
                }
            });
        } catch (RejectedExecutionException e) {
            // 【防御性编程】线程池拒绝执行（已满或已关闭）时的兜底处理：
            // - lambda 不会被执行，因此 lambda 内的 release(msg) 永远不会被调用
            // - 如果不在此处释放，msg.retain() 增加的引用计数将永远无法归零 → 内存泄漏
            //
            // 注意：先记录日志再释放，因为 release 后底层内存可能立即被回收复用，
            // 再访问 msg.variableHeader().topicName() 属于访问已释放资源（不安全）
            LOG.warn("ThreadPool rejected PUBLISH notification for topic={}, clientId={}",
                    msg.variableHeader().topicName(), clientID, e);
            ReferenceCountUtil.release(msg);
            // 重新抛出异常，让调用方感知线程池拒绝事件
            throw e;
        }
    }

**Reply by andsel** [reply]
Hi @AAA123369 thanks for you comment, but the language used to interact in this project is English. So please 🙏 switch to that language for next comments. This GH discussion is used to know how clients of Moquette broker use it.

**Comment by AAA123369** [comment]
/**
 * Notification interceptor: triggered when a client publishes an MQTT message
 * <p>
 * 【ByteBuf Reference Counting Management】
 * Netty's ByteBuf uses reference counting (refCount) to manage the lifecycle of underlying direct memory.
 * - refCount > 0: memory is valid and can be safely accessed
 * - refCount = 0: memory has been released; accessing it will cause a UseCntException
 * - GC only reclaims Java objects (ByteBuf shells), but does NOT call release(), so underlying direct memory is not returned to the pool
 * - release() must be called explicitly to return memory to the Netty memory pool
 * <p>
 * 【Reference Count Flow in This Method】
 * 1. The caller passes in msg with refCount = N (managed by the Netty Pipeline)
 * 2. msg.retain(): refCount = N+1, ensuring memory is not released by the Pipeline during asynchronous execution
 * 3. Each handler receives msg.retainedDuplicate(): creates an independent copy (refCount = N+2), must call release() after use
 * 4. release(msg) before lambda ends: refCount returns to N+1 → subsequent Pipeline release returns to N
 * <p>
 * 【Fix Description - Moquette 0.17 ByteBuf Leak Defect】
 * Original code: handler.onPublish(new InterceptPublishMessage(msg.retainedDuplicate(), ...))
 * Problem: The retainedDuplicate() copy was never release()'d, causing a ByteBuf direct memory leak for every message processed.
 * Fix: Ensure the copy is released via a finally block after handler processing completes.
 *
 * @param msg      MQTT publish message (contains topic, payload, etc., with underlying reference-counted ByteBuf)
 * @param clientID ID of the client publishing the message
 * @param username Username of the publisher
 */
@Override
public void notifyTopicPublished(final MqttPublishMessage msg, final String clientID, final String username) {
    // Increase reference count to prevent the message from being released by the Netty Pipeline during async execution
    // After retain(), refCount = N+1; the corresponding release() is in the lambda's finally block
    msg.retain();

    try {
        // Submit the interceptor notification task to the thread pool for async execution
        executor.execute(() -> {
            try {
                // Extract message metadata (topic and messageId are String/int types, no ByteBuf references)
                int messageId = msg.variableHeader().messageId();
                String topic = msg.variableHeader().topicName();

                // Iterate over all interceptors registered for PUBLISH events and notify each
                for (InterceptHandler handler : handlers.get(InterceptPublishMessage.class)) {
                    LOG.debug("Notifying MQTT PUBLISH message to interceptor. CId={}, messageId={}, topic={}, "
                            + "interceptorId={}", clientID, messageId, topic, handler.getID());

                    // retainedDuplicate() creates an independent copy sharing the underlying data:
                    // - The copy has its own independent readerIndex/writerIndex
                    // - The underlying ByteBuf's refCount is incremented by 1 (so release() is mandatory after use)
                    MqttPublishMessage dup = msg.retainedDuplicate();
                    try {
                        // Pass the copy to the interceptor for processing
                        // The interceptor may read message content but should NOT hold or delay the release of this copy
                        handler.onPublish(new InterceptPublishMessage(dup, clientID, username));
                    } finally {
                        // 【Critical Fix】Release the copy's direct memory regardless of whether the handler throws an exception
                        // Failure to release will cause ByteBuf leaks: GC only reclaims Java objects, but underlying memory is never returned to the pool
                        ReferenceCountUtil.release(dup);
                    }
                }
            } finally {
                // Release the reference count incremented by retain(), paired with msg.retain() at line 125
                // At this point, refCount decreases from N+1 back to N, where the Netty Pipeline will manage it normally thereafter
                ReferenceCountUtil.release(msg);
            }
        });
    } catch (RejectedExecutionException e) {
        // 【Defensive Programming】Fallback handling when the thread pool rejects execution (full or shut down):
        // - The lambda will not execute, so release(msg) inside the lambda will NEVER be called
        // - Without releasing here, the reference count incremented by msg.retain() would never go to zero → memory leak
        //
        // Note: Log is recorded BEFORE release because after release, the underlying memory may be immediately reclaimed and reused,
        // and accessing msg.variableHeader().topicName() afterward would be accessing already-released resources (unsafe)
        LOG.warn("ThreadPool rejected PUBLISH notification for topic={}, clientId={}",
                msg.variableHeader().topicName(), clientID, e);
        ReferenceCountUtil.release(msg);
        // Re-throw the exception to let the caller be aware of the thread pool rejection event
        throw e;
    }
}
